Main/Source/game.cpp

```diff
@@ -3270,50 +3277,93 @@ int game::Load(cfestring& saveName)
+  if(i!=std::string::npos){
+    if(bAlsoFixIt){
+      fs.Empty();
+      fs<<strChk.substr(0,i).c_str();DBG1(fs.CStr());
+    }
+    return true;
+  }
+
+  return false;
+}
+
+festring game::SaveName(cfestring& Base,bool bLoadingFromAnAutosave)
+{
+  festring PathAndBaseSaveName = GetSaveDir();
 
   /**
    * Base must come OK, will just prepend directory,
    * the problem on modifying it is that as it is read from the filesystem
    * it will not be found if it gets changed...
    */
-  DBG3(PlayerName.CStr(), Base.CStr(), AutoSaveFileName.CStr());
+  DBG3(PlayerName.CStr(), Base.CStr(), CurrentBaseSaveFileName.CStr());
 
   if(Base.GetSize() > 0)
   {
-    AutoSaveFileName.Empty();
-    AutoSaveFileName << Base;
-    SaveName << Base;
+    CurrentBaseSaveFileName.Empty();
+    CurrentBaseSaveFileName << Base;
+    chkAutoSaveSuffix(CurrentBaseSaveFileName,true);
+
+    PathAndBaseSaveName << Base;
   }
   else
   {
-    if(AutoSaveFileName.GetSize() == 0)
+    // this is important in case player name changes like when using the fantasy name generator
+    festring fsPN; fsPN<<PlayerName; fixChars(fsPN);
+    std::string strASFN; strASFN = CurrentBaseSaveFileName.CStr();
+    if(strASFN.substr(0,fsPN.GetSize()) != fsPN.CStr())
+      CurrentBaseSaveFileName.Empty();
+
+    if(CurrentBaseSaveFileName.GetSize() == 0)
     {
-      AutoSaveFileName << PlayerName << '_' << time(0);  // e.g. PlayerName-1529392480
-  
-      for(festring::sizetype i = 0; i < AutoSaveFileName.GetSize(); ++i)
-      {
-        // this prevents all possibly troublesome characters in all OSs
-        if(AutoSaveFileName[i]>='A' && AutoSaveFileName[i]<='Z')continue;
-        if(AutoSaveFileName[i]>='a' && AutoSaveFileName[i]<='z')continue;
-        if(AutoSaveFileName[i]>='0' && AutoSaveFileName[i]<='9')continue;
-  
-        AutoSaveFileName[i] = '_';
-      }
+      int iTmSz=100; char cTime[iTmSz]; time_t now = time(0);
+      strftime(cTime,iTmSz,"%Y%m%d_%H%M%S",localtime(&now)); //pretty DtTm
+
+      CurrentBaseSaveFileName << PlayerName << '_' << cTime;
+      fixChars(CurrentBaseSaveFileName);
     }
-    SaveName << AutoSaveFileName;
+    
+    PathAndBaseSaveName << CurrentBaseSaveFileName;
   }
 
-  DBG4(PlayerName.CStr(), SaveName.CStr(), Base.CStr(), AutoSaveFileName.CStr());
+  DBG4(PlayerName.CStr(), PathAndBaseSaveName.CStr(), Base.CStr(), CurrentBaseSaveFileName.CStr());
 
 #if defined(__DJGPP__)
-  if(SaveName.GetSize() > 13)
-    SaveName.Resize(13);
+  if(PathAndBaseSaveName.GetSize() > 13)
+    PathAndBaseSaveName.Resize(13);
 #endif
 
-  return SaveName;
+  if(!bLoadingFromAnAutosave){ //very specific use case
+    if(chkAutoSaveSuffix(PathAndBaseSaveName)){
+      /**
+       * this ABORT is important to prevent the troubling autosave suffix duplicity,
+       * it's consistency is kept using only: GetAutoSaveFileName()
+       */
+      ABORT("The base savegame filename '%s' must not contain '%s'",PathAndBaseSaveName.CStr(),AUTOSAVE_SUFFIX);
```

Comment:
"must not *end with*"?

Reply:
some funny user can create his player name like ".AutoSave funny guy", better avoid it if possible :>,
a better way would be to have more checks, so or it or more constraints :)